pkg/network/kube-proxy.go

```diff
@@ -0,0 +1,175 @@
+package network
+
+import (
+	"log"
+	"net"
+	"os"
+	"path/filepath"
+	"sort"
+
+	"github.com/pkg/errors"
+	uns "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	operv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-network-operator/pkg/render"
+)
+
+// ShouldDeployKubeProxy determines if the desired network type should
+// install kube-proxy.
+// openshift-sdn and OVN-Kubernetes deploy their own kube-proxy. All other
+// network providers are assumed to require kube-proxy
+func ShouldDeployKubeProxy(conf *operv1.NetworkSpec) bool {
+	switch conf.DefaultNetwork.Type {
+	case operv1.NetworkTypeOpenShiftSDN:
+		return false
+	case operv1.NetworkTypeOVNKubernetes:
+		return false
+	default:
+		log.Printf("NOTICE: Unknown network type %s, installing kube-proxy", conf.DefaultNetwork.Type)
+		return true
+	}
+}
+
+// ValidateKubeProxy checks that the kube-proxy configuration is valid.
+// This is checked even if DeployKubeProxy is false, since it also affects
+// openshift-sdn.
+func ValidateKubeProxy(conf *operv1.NetworkSpec) []error {
+	if conf.KubeProxyConfig == nil {
+		return nil
+	}
+
+	out := []error{}
+	kpc := conf.KubeProxyConfig
+	if kpc.BindAddress != "" {
+		ip := net.ParseIP(kpc.BindAddress)
+		if ip == nil {
+			out = append(out, errors.Errorf("Invalid Spec.KubeProxyConfig.BindAddress %q", kpc.BindAddress))
+		}
+	}
+
+	return out
+}
+
+// IsKubeProxyChangeSafe is noop, but it would check if the proposed kube-proxy
+// change is safe.
+func IsKubeProxyChangeSafe(prev, next *operv1.NetworkSpec) []error {
+	// At present, all kube-proxy changes are safe to deploy
+	return nil
+}
+
+// FillKubeProxyDefaults inserts kube-proxy defaults, but only if
+// kube-proxy will be deployed explicitly.
+func FillKubeProxyDefaults(conf, previous *operv1.NetworkSpec) {
+	if conf.DeployKubeProxy == nil {
+		v := ShouldDeployKubeProxy(conf)
+		conf.DeployKubeProxy = &v
+	}
+
+	if !*conf.DeployKubeProxy {
+		return
+	}
+
+	if conf.KubeProxyConfig == nil {
+		conf.KubeProxyConfig = &operv1.ProxyConfig{}
+	}
+
+	if conf.KubeProxyConfig.BindAddress == "" {
+		conf.KubeProxyConfig.BindAddress = "0.0.0.0"
+	}
+
+	if conf.KubeProxyConfig.ProxyArguments == nil {
+		conf.KubeProxyConfig.ProxyArguments = map[string][]string{}
+	}
+
+	if conf.KubeProxyConfig.ProxyArguments["metrics-bind-address"] == nil {
+		conf.KubeProxyConfig.ProxyArguments["metrics-bind-address"] = []string{"0.0.0.0"}
+	}
+
+	if conf.KubeProxyConfig.ProxyArguments["metrics-port"] == nil {
+		conf.KubeProxyConfig.ProxyArguments["metrics-port"] = []string{"9101"}
```

Comment:
Note that this isn't the upstream default. I'm not sure why we override it in openshift-sdn.

Reply:
Because this is accessed remotely (from Prometheus). So, it needs to be in the open port range (9000-9999)

Reply:
Because this is accessed remotely (from Prometheus). So, it needs to be in the open port range (9000-9999)

Reply:
Because this is accessed remotely (from Prometheus). So, it needs to be in the open port range (9000-9999)